### PR TITLE
sql: anonymize function names for error reporting and telemetry

### PR DESCRIPTION
**Pull request by rharding6373**
Previously, we did not anonymize/redact function names because we only supported built-in, CRDB-named functions. However, since adding support for UDFs, function names may be user-provided. We need to omit user-provided names in error reporting and telemetry.

We fix this oversight in this PR by no longer overriding the anonymizing format flag and redaction flag for FuncExprs. Unfortunately, since any functions in a query are not always resolved when the error occurs, we may also redact built-in function names.

Epic: None
Fixes: None

Release note (sql change): Anonymizes function names in error reporting and telemetry so that user-provided names for UDFs are not included.

### CODE REVIEW

**Review by rharding6373 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110870)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @mgartner)

___
*[`pkg/sql/sem/tree/expr.go` line 1378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NefdNV48ARUzmtX_qEH:-NefdNV48ARUzmtX_qEI:bxn02pm) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/sem/tree/expr.go#L1378)):*
> ```Go
> 
> 	// Ideally we would not anonymize/redact built-in function names. Some
> 	// built-in functions are correctly typed as FunctionDefinition and so avoid
> ```

@ecwall maybe you can help answer this question for me. `FunctionDefinition` has a [comment](https://github.com/cockroachdb/cockroach/blob/ef0bc41022d0241ea5bcaff2c046b7f798cbe212/pkg/sql/sem/tree/function_definition.go#L25-L26) that states that it refers to only overloads for built-in functions. With UDFs implemented, is this still true? If yes, then we get some non-anonymized built-ins (though judging by the tests, it's not reliable). If not, I can make sure all potential function names are redacted by adding some logic handling `FmtAnonymize` in `FunctionDefinition`'s `Format()`.

**Review by mgartner — COMMENTED** [review]
> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110870)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rharding6373)

___
*[`pkg/sql/sem/tree/expr.go` line 1378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NefdNV48ARUzmtX_qEH:-NehlPsp81jGRzq4dhLh:b-avi5ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/sem/tree/expr.go#L1378)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

@ecwall maybe you can help answer this question for me. `FunctionDefinition` has a [comment](https://github.com/cockroachdb/cockroach/blob/ef0bc41022d0241ea5bcaff2c046b7f798cbe212/pkg/sql/sem/tree/function_definition.go#L25-L26) that states that it refers to only overloads for built-in functions. With UDFs implemented, is this still true? If yes, then we get some non-anonymized built-ins (though judging by the tests, it's not reliable). If not, I can make sure all potential function names are redacted by adding some logic handling `FmtAnonymize` in `FunctionDefinition`'s `Format()`.
</blockquote></details>

I'm not sure if that's true, but even if true, a user-defined function could have the same name as a built-in and take precedence during resolution.

nit: Can you explain why built-in function names must be redacted in the comment? It'll be helpful if someone stumbles upon this in the future.
___
*[`pkg/sql/parser/testdata/user_defined_functions` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NehmWFnCqCWWhCzE8wl:-NehmWFnCqCWWhCzE8wm:bf7nh7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/parser/testdata/user_defined_functions#L31)):*
> ```
> SELECT (substring((udf((1), (2))), (3))) -- fully parenthesized
> SELECT substring(udf(_, _), _) -- literals removed
> SELECT substring(_(1, 2), 3) -- identifiers removed
> ```

Hmm - I presume this is one of the inconsistent cases you mention in your other comment? I wonder what makes `substring` special compared to other built-ins like `now`.

**Review by rharding6373 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110870)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @mgartner)

___
*[`pkg/sql/parser/testdata/user_defined_functions` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NehmWFnCqCWWhCzE8wl:-NeiCESvEDaEebJtV7o6:b-p0zui9) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/parser/testdata/user_defined_functions#L31)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Hmm - I presume this is one of the inconsistent cases you mention in your other comment? I wonder what makes `substring` special compared to other built-ins like `now`.
</blockquote></details>

You might be intrigued by my experiment with overloading a couple different builtins: https://gist.github.com/rharding6373/d456102f4fb2b9d21724ad184318ef25

But yes, this is one of the tests that led to my question above. If you look at the predefined_functions test,  most built-in functions are identified as FunctionDefinitions and are not redacted, but any that are called as arguments are not.

**Review by mgartner — APPROVED** [review]
Thanks for taking care of this!

:lgtm: modulo the nit about expanding on the comment, if you want to address that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-:-NeixFGf1Dn2oU9qh4v6:b-zbut7y)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ecwall and @rharding6373)

___
*[`pkg/sql/parser/testdata/user_defined_functions` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NehmWFnCqCWWhCzE8wl:-Neiv4WB6HLL6oHcVq3M:b-j16hi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/parser/testdata/user_defined_functions#L31)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

You might be intrigued by my experiment with overloading a couple different builtins: https://gist.github.com/rharding6373/d456102f4fb2b9d21724ad184318ef25

But yes, this is one of the tests that led to my question above. If you look at the predefined_functions test,  most built-in functions are identified as FunctionDefinitions and are not redacted, but any that are called as arguments are not.
</blockquote></details>

Your substring example is interesting. It looks like it can never resolve as a UDF because it's immediately resolved to a built-in via the parser here: https://github.com/cockroachdb/cockroach/blob/b9f4a2d2a29d5d010c82d8c1e45296c0ccd55a3a/pkg/sql/parser/sql.y#L15301

That's why it's a FunctionDefinition rather than a FuncExpr, and isn't anonymized, I think. I believe it'll happen with any of those keyword functions.

Postgres doesn't let you create a function named `substring`, `trim`,  `overlay`, or any other of the "keyword functions" defined in the grammar.:

```
marcus=# CREATE FUNCTION substring() RETURNS VOID LANGUAGE SQL AS '';
ERROR:  42601: syntax error at or near "("
LINE 1: CREATE FUNCTION substring() RETURNS VOID LANGUAGE SQL AS '';
                                 ^
LOCATION:  scanner_yyerror, scan.l:1180
```

We should probably have the same behavior here.

**Comment by mgartner** [comment]
*[`pkg/sql/parser/testdata/user_defined_functions` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NehmWFnCqCWWhCzE8wl:-NeiykMf5RFlOlMnqdri:b-ya5xd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/parser/testdata/user_defined_functions#L31)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Your substring example is interesting. It looks like it can never resolve as a UDF because it's immediately resolved to a built-in via the parser here: https://github.com/cockroachdb/cockroach/blob/b9f4a2d2a29d5d010c82d8c1e45296c0ccd55a3a/pkg/sql/parser/sql.y#L15301

That's why it's a FunctionDefinition rather than a FuncExpr, and isn't anonymized, I think. I believe it'll happen with any of those keyword functions.

Postgres doesn't let you create a function named `substring`, `trim`,  `overlay`, or any other of the "keyword functions" defined in the grammar.:

```
marcus=# CREATE FUNCTION substring() RETURNS VOID LANGUAGE SQL AS '';
ERROR:  42601: syntax error at or near "("
LINE 1: CREATE FUNCTION substring() RETURNS VOID LANGUAGE SQL AS '';
                                 ^
LOCATION:  scanner_yyerror, scan.l:1180
```

We should probably have the same behavior here.
</blockquote></details>

To summarize, I think we could keep [the builtin functions with keyword names](https://github.com/cockroachdb/cockroach/blob/b9f4a2d2a29d5d010c82d8c1e45296c0ccd55a3a/pkg/sql/parser/sql.y#L15185-L15330) non-anonymous, but we'd have to prevent creating UDFs with those names (which we should do anyway for PG compatibility and because those UDFs don't currently work - you can never execute them).

**Comment by rafiss** [comment]
it seems like we should be able to use a type assertion to get the `ResolvedFunctionDefinition` or `FunctionDefinition`, and see which schema the function belongs too. if it's `pg_catalog`, `crdb_internal` or `information_schema`, we know it's a builtin.

**Comment by mgartner** [comment]
This is an anonymization of the AST which may or may-not be type-checked at the point of anonymization, so it won't necessarily be a ResolvedFunctionDefinition. It might just be a FuncExpr.

File: pkg/sql/sem/tree/expr.go [hunk 1]
```diff
@@ -1374,11 +1374,10 @@ func (node *FuncExpr) Format(ctx *FmtCtx) {
...
 
-	// We need to remove name anonymization/redaction for the function name in
-	// particular. Do this by overriding the flags.
-	// TODO(thomas): when function names are correctly typed as FunctionDefinition
-	// remove FmtMarkRedactionNode from being overridden.
-	ctx.WithFlags(ctx.flags&^FmtAnonymize&^FmtMarkRedactionNode|FmtBareIdentifiers, func() {
+	// Ideally we would not anonymize/redact built-in function names. Some
+	// built-in functions are correctly typed as FunctionDefinition and so avoid
+	// redaction.
+	ctx.WithFlags(ctx.flags|FmtBareIdentifiers, func() {
```
Comment (rafiss) [inline pkg/sql/sem/tree/expr.go:1380]: not blocking this PR, but i wonder if it's true that `node.fnProps` is only non-nil for builtin functions. that could be a way to determine if the name needs to be redacted

Comment (mgartner) [reply pkg/sql/sem/tree/expr.go:1383]: `fnProps` is written to during type-checking when the overload is resolved. At that point we can determine if the function is built-in or not, using `Overload.Type == BuiltinRoutine`. But if the AST is printed before being type-checked, we cannot.

**Review by rharding6373 — COMMENTED** [review]
TFTRs! RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-:-NfOnfcZ8JdrKHQy1Nio:bk4fma)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ecwall, @mgartner, and @rafiss)

___
*[`pkg/sql/sem/tree/expr.go` line 1378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NefdNV48ARUzmtX_qEH:-NfNYWh1-jneCHBBr3YE:bm3qjnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/sem/tree/expr.go#L1378)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not sure if that's true, but even if true, a user-defined function could have the same name as a built-in and take precedence during resolution.

nit: Can you explain why built-in function names must be redacted in the comment? It'll be helpful if someone stumbles upon this in the future.
</blockquote></details>

I did more digging, and FunctionDefinition is only used for built-ins except for sqlsmith, which uses this type to track UDFs. Therefore I kept anonymization out of FunctionDefinition.Format() and ResolvedFunctionDefinition.Format(), but added comments about the assumption that these are only for built-ins.

Unfortunately this makes anonymization and redaction inconsistent (i.e., redacted if formatted before resolution, not redacted after), but this currently seems like a reasonable compromise so we don't need to redact the function names all the time. Let me know what you think.
___
*[`pkg/sql/sem/tree/expr.go` line 1380 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#gh-1333555256:-NfNZDRG5Pil18ve05hP:btnjbhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/sem/tree/expr.go#L1380)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`fnProps` is written to during type-checking when the overload is resolved. At that point we can determine if the function is built-in or not, using `Overload.Type == BuiltinRoutine`. But if the AST is printed before being type-checked, we cannot.
</blockquote></details>

I opened https://github.com/cockroachdb/cockroach/issues/111385 for further investigation into identifying builtins. Let's use the other comment thread to determine whether this PR does enough to try to preserve builtins, so the discussion is in one place.
___
*[`pkg/sql/parser/testdata/user_defined_functions` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-NehmWFnCqCWWhCzE8wl:-NfNfQjM4lxT3__Hwt2X:b-tnhcru) ([raw file](https://github.com/cockroachdb/cockroach/blob/25237f4768059b1d30da217c42ff780d923854e3/pkg/sql/parser/testdata/user_defined_functions#L31)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

To summarize, I think we could keep [the builtin functions with keyword names](https://github.com/cockroachdb/cockroach/blob/b9f4a2d2a29d5d010c82d8c1e45296c0ccd55a3a/pkg/sql/parser/sql.y#L15185-L15330) non-anonymous, but we'd have to prevent creating UDFs with those names (which we should do anyway for PG compatibility and because those UDFs don't currently work - you can never execute them).
</blockquote></details>

I will open a separate PR to error in the parser, since that's orthogonal to this PR. 

With the current changes, the built-in keywords should not be redacted.

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110870)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ecwall, @mgartner, and @rharding6373)

___
*[`pkg/sql/statement_mark_redaction_test.go` line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-Nfv_VeN9ZTY6B4Rz09c:-Nfv_VeN9ZTY6B4Rz09d:b-wngv4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8a658c7249be326ad5de829e044cb5aa384884/pkg/sql/statement_mark_redaction_test.go#L53)):*
> ```Go
> 		{
> 			"SELECT crdb_internal.node_executable_version()",
> 			"SELECT ‹crdb_internal›.‹node_executable_version›()",
> ```

one more thought - i believe we can safely not redact any function whose name begins with `crdb_internal.` (including that last `.`). if you'd like to leave that for https://github.com/cockroachdb/cockroach/issues/111385, that's fine with me, but it does seem useful to know if an error report is caused by a crdb_internal function, since we probably would prioritize those differently.

**Review by rharding6373 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110870)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ecwall, @mgartner, and @rafiss)

___
*[`pkg/sql/statement_mark_redaction_test.go` line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110870#-Nfv_VeN9ZTY6B4Rz09c:-Nfww5GpENrGp_4Uq4CT:bade5he) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8a658c7249be326ad5de829e044cb5aa384884/pkg/sql/statement_mark_redaction_test.go#L53)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

one more thought - i believe we can safely not redact any function whose name begins with `crdb_internal.` (including that last `.`). if you'd like to leave that for https://github.com/cockroachdb/cockroach/issues/111385, that's fine with me, but it does seem useful to know if an error report is caused by a crdb_internal function, since we probably would prioritize those differently.
</blockquote></details>

Agree that it would be nice not to redact it. I figured out where to check for that, so we should no longer redact any unresolved names that start with `crdb_internal.`.

**Comment by rharding6373** [comment]
TFTRs!

bors r+
